### DISCUSSION TITLE
How is hot-swapping of matching databases implemented?

### DISCUSSION THREAD

**Discussion by benbelow** [Anthony-Nolan/Atlas#788] [Q&A] [answered]
From @daria-sorokina-da: "Databases A, B - when does hot-swapping happen? how is it implemented?"

**Comment by zabeen** [comment, marked as answer]
## When does hot-swapping happen?

When the [data refresh job](https://github.com/Anthony-Nolan/Atlas/blob/master/README_Integration.md#data-refresh) completes successfully. 

The idea is that the newly processed data should not be accessed for search until the job to re-process all donors with the up-to-date nomenclature version has been fully completed.

## How is it implemented?

In the "MatchingAlgorithmPersistent" schema of the shared Atlas database, there is a table containing details of all historic data refresh records. 

This is not just an audit for support purposes, but provides the definitive source of which database to use: 

* `The database against which the most recent successful data refresh was run should be used for search`

As part of the data refresh, a row will be added to this table, gradually updated with details on how far through it is, and eventually marked as a success via a boolean column. All in flight searches will finish against the same database they started with (thanks to per-request caching of the "active database" state) - and all new searches will begin against the newly active database.
